File: examples/typescript/src/workflow/datasetdeal/actions.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { StateData } from "./models.js";

export const DatasetDealAction = Object.freeze({
  TRANSFER_MONEY_FROM_BUYER_TO_SELLER: "transferMoneyFromBuyerToSeller",
  TRANSFER_MONEY_FROM_SELLER_TO_BUYER: "transferMoneyFromSellerToBuyer",
  TRANSPORT_FULL_DATASET_TO_BUYER: "transportFullDatasetToBuyer",
  TRANSPORT_SAMPLE_DATASET_TO_BUYER: "transportSampleDatasetToBuyer",
} as const);

export interface ActionInput {
  readonly flowId: string;
  readonly processId: string;
  readonly buyerId: string;
  readonly targetState: string;
  readonly stateData: StateData;
}

type ActionHandler = (input: ActionInput) => StateData | Promise<StateData>;

export class DatasetDealActions {
  private readonly handlers: ReadonlyMap<string, ActionHandler>;

  public constructor() {
    this.handlers = new Map<string, ActionHandler>([
      [
        DatasetDealAction.TRANSFER_MONEY_FROM_BUYER_TO_SELLER,
        transferMoneyFromBuyerToSeller,
      ],
      [
        DatasetDealAction.TRANSFER_MONEY_FROM_SELLER_TO_BUYER,
        transferMoneyFromSellerToBuyer,
      ],
      [DatasetDealAction.TRANSPORT_FULL_DATASET_TO_BUYER, transportFullDatasetToBuyer],
      [DatasetDealAction.TRANSPORT_SAMPLE_DATASET_TO_BUYER, transportSampleDatasetToBuyer],
    ]);
  }

  public availableNames(): readonly string[] {
    return [...this.handlers.keys()].sort();
  }

  public async execute(name: string, input: ActionInput): Promise<StateData> {
    const handler = this.handlers.get(name);
    if (handler === undefined) {
      throw new TypeError(`dataset deal action ${name} is not registered`);
    }
    return {
      ...(await handler(input)),
      lastAction: name,
      lastActionStatus: "completed",
    };
  }
}

function transferMoneyFromBuyerToSeller(input: ActionInput): StateData {
  console.info("dataset deal transferred money from buyer to seller", {
    flowId: input.flowId,
    buyerId: input.buyerId,
    targetState: input.targetState,
    samplePrice: input.stateData.proposedSamplePrice,
    fullPrice: input.stateData.proposedFullPrice,
  });
  return {};
}

function transferMoneyFromSellerToBuyer(input: ActionInput): StateData {
  console.info("dataset deal transferred refund from seller to buyer", {
    flowId: input.flowId,
    buyerId: input.buyerId,
    refundPrice: input.stateData.proposedSampleRefundPrice,
  });
  return {};
}

function transportFullDatasetToBuyer(input: ActionInput): StateData {
  console.info("dataset deal transported full dataset to buyer", {
    flowId: input.flowId,
    buyerId: input.buyerId,
  });
  return { deliveredDataset: "full" };
}

function transportSampleDatasetToBuyer(input: ActionInput): StateData {
  console.info("dataset deal transported sample dataset to buyer", {
    flowId: input.flowId,
    buyerId: input.buyerId,
  });
  return { deliveredDataset: "sample" };
}

File: examples/go/products/dataset-deal/actions.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package datasetdeal

import (
	"fmt"
	"sort"

	"github.com/superdurable/dex/sdk-go/dex"
	"github.com/superdurable/dex/sdk-go/logging"
)

const (
	TransferMoneyFromBuyerToSeller = "transferMoneyFromBuyerToSeller"
	TransferMoneyFromSellerToBuyer = "transferMoneyFromSellerToBuyer"
	TransportFullDatasetToBuyer    = "transportFullDatasetToBuyer"
	TransportSampleDatasetToBuyer  = "transportSampleDatasetToBuyer"
)

type ActionInput struct {
	FlowID      string
	ProcessID   string
	BuyerID     string
	TargetState string
	StateData   map[string]string
}

type actionHandler func(dex.Logger, ActionInput) (map[string]string, error)

type actionRegistry struct {
	logger   dex.Logger
	handlers map[string]actionHandler
}

func newActionRegistry(logger dex.Logger) *actionRegistry {
	return &actionRegistry{
		logger: logging.OrDefault(logger),
		handlers: map[string]actionHandler{
			TransferMoneyFromBuyerToSeller: transferMoneyFromBuyerToSeller,
			TransferMoneyFromSellerToBuyer: transferMoneyFromSellerToBuyer,
			TransportFullDatasetToBuyer:    transportFullDatasetToBuyer,
			TransportSampleDatasetToBuyer:  transportSampleDatasetToBuyer,
		},
	}
}

func (registry *actionRegistry) execute(
	name string,
	input ActionInput,
) (map[string]string, error) {
	handler, found := registry.handlers[name]
	if !found {
		return nil, fmt.Errorf("dataset deal action %q is not registered", name)
	}
	updates, err := handler(registry.logger, input)
	if err != nil {
		return nil, err
	}
	if updates == nil {
		updates = make(map[string]string)
	}
	updates["lastAction"] = name
	updates["lastActionStatus"] = "completed"
	return updates, nil
}

func AvailableActionNames() []string {
	names := []string{
		TransferMoneyFromBuyerToSeller,
		TransferMoneyFromSellerToBuyer,
		TransportFullDatasetToBuyer,
		TransportSampleDatasetToBuyer,
	}
	sort.Strings(names)
	return names
}

func transferMoneyFromBuyerToSeller(
	logger dex.Logger,
	input ActionInput,
) (map[string]string, error) {
	logger.Info(
		"dataset deal transferred money from buyer to seller",
		"flow_id", input.FlowID,
		"buyer_id", input.BuyerID,
		"target_state", input.TargetState,
		"sample_price", input.StateData["proposedSamplePrice"],
		"full_price", input.StateData["proposedFullPrice"],
	)
	return nil, nil
}

func transferMoneyFromSellerToBuyer(
	logger dex.Logger,
	input ActionInput,
) (map[string]string, error) {
	logger.Info(
		"dataset deal transferred refund from seller to buyer",
		"flow_id", input.FlowID,
		"buyer_id", input.BuyerID,
		"refund_price", input.StateData["proposedSampleRefundPrice"],
	)
	return nil, nil
}

func transportFullDatasetToBuyer(
	logger dex.Logger,
	input ActionInput,
) (map[string]string, error) {
	logger.Info(
		"dataset deal transported full dataset to buyer",
		"flow_id", input.FlowID,
		"buyer_id", input.BuyerID,
	)
	return map[string]string{"deliveredDataset": "full"}, nil
}

func transportSampleDatasetToBuyer(
	logger dex.Logger,
	input ActionInput,
) (map[string]string, error) {
	logger.Info(
		"dataset deal transported sample dataset to buyer",
		"flow_id", input.FlowID,
		"buyer_id", input.BuyerID,
	)
	return map[string]string{"deliveredDataset": "sample"}, nil
}
